### ISSUE TITLE
Test for mariadb multi-AZ ACID properties during partial failures

### ISSUE THREAD

**Issue by gberche-orange** [orange-cloudfoundry/paas-templates#965]
### Expected behavior

* As a marketplace user 
* In order to ensure my app using dedicated mariadb multi-AZ #612 meeting its SLA
* I need that [ACID](https://en.wikipedia.org/wiki/ACID) properties be garanteed during partial failure modes

* As a paas-templates operator 
* In order to be confident that ACID properties are fulfilled during partial outages
* I need (automated) tests that test ACID properties during partial outages

### Observed behavior
 
The following scenarios need to be explicitly tested

Given a client configured to fail over between proxies in r1 and r2
* Given r1 to r2 communication is down, but r1 to r3 and r1 to r2 communication is up
* When client containers in r1 keep retrying commit `a INSERT with primary key k` transaction T1 to the database (using r1 proxy)
* When client containers in r2 keep retrying commit `a INSERT with primary key k`  transaction T2 (using r2 proxy)
* Then exactly one of T1 or T2 gets commited

Given a client configured to fail over between proxies in r1 and r2
* Given r1 to r2 communication is down, but r1 to r3 and r1 to r2 communication is up
* When client containers in r1 constantly commit `a INSERT with new primary key` transaction T1 to the database (using r1 proxy)
* When client containers in r2 constantly commit `a INSERT with new primary key`  transaction T2 (using r2 proxy)
* Then exactly T1 and T2 gets commited, 
* And all inserted rows are visible to with r1 and r2

/CC @CDallosto-orange @ldangeard-orange @poblin-orange 

### Affected releases

* 48.x
* earlier versions

**Comment by ldangeard-orange** [comment]
ACID ownership is provided by Galera Cluster, regardless of the number of nodes, multi-az, with or without writes on 1 single node or several nodes, of a Bosh, K8s deployment or on VMS ...

Also in the Bosh release,
  - we must not use the 2 proxies (see doc release),
  - potentially the proxies redirect transactions to the same node.

Other points:
  - it is the COMMIT which checks that the other nodes apply the transaction (checking of DEAD LOCK)
- the developer must therefore ensure the COMMIT return code

**Comment by poblin-orange** [comment]
So the test should be targeted to both bosh release proxies used concurrently, due to r1-r2 network split ?
And the risk we have to assert is deadlocked transactions ?

**Comment by gcordierOrange** [comment]
I made this test 
- R1: app => Proxy 1 => mysql 2 (OK)
- R2: app => Proxy 2 => mysql 2 (OK)

The mysql node is elected by the proxy, in normal mode, the both proxy elect the same node (min wsrep_local_index).

I break the link between R1/R2 (R1/R3 and R2/R3 are still up)
Galera detect no lost off node (cluster_size = 3), but switchboard change galera node for R2 (because off healcheck for proxy1 is ko for mysql2 and only mysql1 is responding)

> curl http://192.168.65.23:9200/api/v1/status
> curl: (7) Failed to connect to 192.168.65.23 port 9200: No route to host

We have this situation 
- R1: app => Proxy 1 => mysql 2 (OK)
- R2: app => Proxy 2 => mysql 1 (OK)
but sometimes : I get a deadlock error

I break the link R1/R2 and R1/R3, only R2/R3 is UP
Galera detect lost off one node node (cluster_size = 2)
- R1 : proxy 1 => ko ( the proxy healcheck test for mysql1 is ko, because mysql1 is on a bad wsrep state, but mysql2 is unreachable, so no galera node are available for proxy)
- R2 : Proxy 2 => mysql 2 (OK) on the same galera node

In conclusion, use the 2 proxy same time is not a good solution.

I also use the mariadb driver with sequential failover, and with the same test, everything is working fine

**Comment by ldangeard-orange** [comment]
Use 1 proxy : 
_"If multiple Switchboard proxies are used in parallel (ex: behind a load-balancer) there is no guarantee that the proxies will choose the same active node. This can result in deadlocks, wherein attempts to update the same row by multiple clients will result one commit succeeding and the other fails. This is a known issue, with exploration of mitigation options on the roadmap for this product. To avoid this problem, use a single proxy instance or an external failover system to direct traffic to one proxy instance at a time."_
https://github.com/orange-cloudfoundry/cf-mysql-release/blob/master/docs/proxy.md

**Comment by gberche-orange** [comment]
> Use 1 proxy :
> "If multiple Switchboard proxies are used in parallel (ex: behind a load-balancer) there is no guarantee that the proxies will choose the same active node. This can result in deadlocks, wherein attempts to update the same row by multiple clients will result one commit succeeding and the other fails. This is a known issue, with exploration of mitigation options on the roadmap for this product. To avoid this problem, use a single proxy instance or an external failover system to direct traffic to one proxy instance at a time."
> See : https://github.com/cloudfoundry/cf-mysql-release/blob/develop/docs/proxy.md
> 

@ldangeard-orange from discussions with cf services lead, this documentation is deprecated/imprecise. Pivotal commercial product is still using multiple proxies behind bosh dns, (not yet clear whether this is fail over mode or load balancing)

See https://github.com/cloudfoundry/cf-mysql-release/blame/develop/docs/proxy.md where consistent routing section was added 4 years after initial disclaimer which has not been removed
![image](https://user-images.githubusercontent.com/4748380/96896160-79b85580-148d-11eb-986d-70f46d13b9c7.png)


See https://github.com/orange-cloudfoundry/paas-templates/issues/716 to associated diagrams of pivotal commercial offering as well as pointers to their documentation
